site/docs/hosting/vps.md

```diff
@@ -50,131 +151,610 @@ cd /etc/systemd/system
 
-#### Restart
+#### Connecting a Webhook to Telegram
 
-By restarting, you stop the app, and start it again.
+All you have to do is tell Telegram where to send the updates.
+To do this, open your browser and visit the page at the following link:
 
-```sh
-pm2 restart <app_name>
+```text
+https://api.telegram.org/bot<token>/setWebhook?url=https://<domain>/<token>
 ```
 
-#### Reload
+> Replace `<token>` with your bot token and `<domain>` with your domain.
+
+## CI/CD
+
+[CI/CD](https://about.gitlab.com/topics/ci-cd) is an important part of the modern software development process.
+This guide covers almost all of the [CI/CD pipeline](https://about.gitlab.com/topics/ci-cd/cicd-pipeline).
+
+We will focus on writing scripts for GitHub and GitLab.
+You can easily adapt the examples below to your CI/CD service of choice, such as Jenkins, Buddy, etc., if needed.
+
+:::tip Self-hosted runner
+GitHub and GitLab offer a certain amount of resources for free to complete your tasks.
+However, when setting up your pipeline, you may quickly use them all up, which will require you to pay money for additional resources or the tasks will not run.
+To prevent this from happening, we recommend installing a self-hosted runner on your computer so that tasks run on your computer.
+This way, you'll get rid of the limits and be able to run pipelines of any complexity and capacity (almost).
+:::
+
+### SSH Keys
 
-By reloading, you replace the current process of your app with a new one, resulting in a 0-second downtime.
-This is recommended for stateless applications.
+To deliver files to the server, you need to set up passwordless authentication, which is implemented using SSH keys.
+
+The following commands should be run on your personal computer.
+
+Change to the directory with the SSH keys:
 
 ```sh
-pm2 reload <app_name>
+cd ~/.ssh
 ```
 
-#### Stop
+Generate a new key pair:
 
-```sh
-# A single app
-pm2 stop <app_name>
+::: code-group
+
+```sh [GitHub]
+ssh-keygen -t rsa -m PEM
+```
 
-# All apps
-pm2 stop all
+```sh [GitLab]
+ssh-keygen -t ed25519
 ```
 
-#### Delete
+:::
+
+This command will generate a public and private key of the type and format you want for GitHub and GitLab.
+You can also specify a custom key name if you wish.
+Note that you should skip the fingerprint generation step (just press `Enter`).
 
-By deleting, you stop your app and remove its logs and metrics.
+Next, send the **public** key to the server:
 
 ```sh
-pm2 del <app_name>
+ssh-copy-id -i <key-name>.pub root@<ip-address>
+```
+
+> Replace `<key-name>` with the name of the generated key and `<ip-address>` with the IP address of your server.
+
+Note that the **public** key can be located on many servers, and the **private** key should be only with you and GitHub or GitLab.
+
+You can now connect to the server without having to enter a password.
+
+### Example Scripts
+
+#### Node.js (GitHub)
+
```

Comment:
```suggestion

Use

```
is necessary because otherwise we have half a sentence with no beginning after the code area